numba/cuda/cudaimpl.py

```diff
@@ -493,6 +493,134 @@ def ptx_fp16_binary(context, builder, sig, args):
+    # it is an approximation and the term 'err' stands for the
+    # relative error of the approximation.
+
+    yn = inline_asm_helper(builder, ir.FloatType(), [ir.FloatType()],
+                           "rcp.approx.ftz.f32 $0, $1;", "=f,f", [y_fp32])
+
+    # Implement fp32 division via multiplication by
+    # reciprocal of arg2
+
+    # Our objective is to get a quotient x/y, correctly rounded to nearest
+    # even in fp16 format, so we can do q0 = x * yn.
+
+    q0 = builder.fmul(x_fp32, yn)
+
+    # Convert fp32 division result to fp16
+
+    fp16_div = inline_asm_helper(builder, ir.IntType(16), [ir.FloatType()],
+                                 "cvt.rn.f16.f32 $0, $1;", "=h,f", [q0])
+
+    # Before performing a second Newton Rapson iteration
+    # filter out small denormal values and zero.
+    # If either case is true skip the second iteration.
+
+    and_cnst = context.get_constant(types.int16, 32767)
+    and_op = builder.and_(fp16_div, and_cnst)
+
+    # Check for small denormals
+    temp_reg = '__$$temp3'
+    inline_asm_helper(builder, ir.VoidType(), [],
+                      f"{{ .reg .pred {temp_reg};", "", [])
+
+    setp_cnst = context.get_constant(types.int16, 143)
+    inline_asm_helper(builder, ir.VoidType(), [ir.IntType(16),
+                                               ir.IntType(16)],
+                      f"setp.lt.f16 {temp_reg}, $0, $1;", "h,h",
+                      [and_op, setp_cnst])
+
+    selp = inline_asm_helper(builder, ir.IntType(16), [],
+                             f"selp.u16 $0, 1, 0, {temp_reg};}}", "=h",
+                             [])
+
+    zero = context.get_constant(types.int16, 0)
+    cmp1 = builder.icmp_unsigned("==",
+                                 builder.bitcast(selp, ir.IntType(16)),
+                                 zero)
+
+    # Check for zero result
+    cmp2 = builder.icmp_unsigned("==",
+                                 builder.bitcast(and_op, ir.IntType(16)),
+                                 zero)
+    # Filter out denormal and precise zero
+    or_op = builder.or_(cmp1, cmp2)
+    or_cond = builder.bitcast(or_op, ir.IntType(1))
+
+    entry_bb = builder.block
+    fall_thru_bb = builder.append_basic_block()
+    true_bb = builder.append_basic_block()
+    builder.cbranch(or_cond, true_bb, fall_thru_bb)
+
+    # Fall through branch (Second Newton Rapson iteration)
+    builder.position_at_start(fall_thru_bb)
+
+    fname = 'llvm.nvvm.fma.rn.f'
+    lmod = builder.module
+    fnty = ir.FunctionType(ir.FloatType(),
+                           (ir.FloatType(),
+                            ir.FloatType(),
+                            ir.FloatType()))
+    fma_func = cgutils.get_or_insert_function(lmod, fnty, fname)
+
+    # We need to perform one more Newton iteration to eliminate the error
+    # from the multiplication.
+    #
+    # Mathematically we have: q0 = x * 1/y * (1 - err)
+    # So err (correction_term) = x - y*q0 = fma(-y, q0, x)
```

Comment:
The variable `err` here means something different to the previous formula. If they mean different thing then it shouldn't share the same name. Probably `correction_term` is better naming.

Reply:
No this is the same 'err' variable. This comment spells out it's derivation. I can drop the use of correction_term to avoid confusion.

Reply:
I think the term on the left is `x * err` (deriving from the first formula). Replacing this with the first term in the next formula:
`x * y_n * err` gives an estimation of the correction term. Adding this to `q0` should bring the result closer to the actual value. Because `q0 = x * 1/y * (1 - err) = x * 1/y - x * 1/y * err`, the second term is the error term that requires correction. We don't have the actual `1/y` but we do have the estimate of it `yn`. Using that to correct should reduce the error by an order of magnitude.

Reply:
I guess this is why it's called an "iteration" of newton's method because the net result of an iteration is reducing the error `epsilon` to `epsilon^2`?

Reply:
Actually @isVoid you were correct, I am about to push another change which should hopefully clear this up.

Reply:
Thanks. I think this should eliminate the ambituity.